### PR TITLE
Check early stopping value under evaluation mode

### PR DESCRIPTION
**Pull request by eywalker**
Currently, the early stopping criteria was evaluated on the validation set (as desired) but on training mode. This is not an appropriate mode for evaluating validation score and also has undesirable side effect of allowing some values such as BathNorm stats to be learned partially on the validation set. This was fixed by ensuring that the validation objective is evaluated on evaluation mode of the model (i.e. after calling `model.eval()`). The model's training status is immediately restored to it's previous value following the evaluation of the objective.

### CODE REVIEW

File: attorch/train.py [hunk 1]
```diff
@@ -31,17 +31,25 @@ def early_stopping(model, objective, interval=5, patience=20, start=0, max_iter=
         max_iter:  maximum number of iterations before the iterator terminated
         maximize:  whether the objective is maximized of minimized
     """
+    training_status = model.training
+
+    def _objective(mod):
```
Comment (fabiansinz) [inline attorch/train.py]: What if, for some reason, some part of the model was set to `train=True` and the other to `train=False`

Comment (eywalker) [inline attorch/train.py]: Hm, that's a rather weird case. Bottom line is that we need the model to be in the evaluation mode (`train(mode=False)`) when performing evaluation on the validation set. I'd prefer to do the switch to the evaluation mode inside the early stopping code to ensure that the user doesn't have to remember doing this switch (as it was forgotten previously). 

In general the model should have two states : `train` vs `eval` and it is really up to the model implementer to override the necessary methods (e.g. `train` and `eval`) if they don't want this status flip to propagate all the way to all submodules.

Comment (eywalker) [inline attorch/train.py]: Actually an easy solution would be to have an optional argument `switch_mode=True` that controls whether more switching will be attempted. If you like prefer the early stopping to not touch the `model`'s training mode, you can set it to `False` 🙂

Comment (fabiansinz) [inline attorch/train.py]: Sounds good.

**Comment by eywalker** [comment]
Made modification based on our discussion. Now `early_stopping` accepts `switch_mode` keyword argument to control whether it will apply `model.eval()` prior to objective evaluation.
